### PR TITLE
Entropy: STM32: add health test configuration

### PR DESCRIPTION
**Pull request by ABOSTM**
Entropy: STM32: add RNG health test configuration

As identified in PR #36993, some STM32 series need to configure health test register
for proper RNG behavior.
In addition, some also require to write a Magic number before writing the configuration.

### CODE REVIEW

**Comment by ABOSTM** [comment]
^^ @fabiobaltieri FYI

File: dts/bindings/rng/st,stm32-rng.yaml
Comment (erwango) [inline dts/bindings/rng/st,stm32-rng.yaml:20]:
```suggestion
    health-test-magic:
```

File: dts/bindings/rng/st,stm32-rng.yaml
Comment (erwango) [inline dts/bindings/rng/st,stm32-rng.yaml:30]:
```suggestion
    health-test-config:
      type: int
      required: false
      description: |
          Health Test confuguration, blabla
```
Please explain when and why this property should be provided.

**Review by erwango — CHANGES_REQUESTED** [review]
Providing more details would be valuable

File: dts/bindings/rng/st,stm32-rng.yaml
Comment (ABOSTM) [reply dts/bindings/rng/st,stm32-rng.yaml:20]: I changed the description, but I would prefer to keep the name `htcfg` as it is the bitfield name used in Reference Manual.
Similarly I would prefer to keep  `htcfg1` as it refer to htcfg from Reference Manual and it is also used in STM32 HAL.
ex:
```
#define RNG_HTCFG_1   0x17590ABCU
#define RNG_HTCFG     0x000CAA74U
```

Comment (erwango) [reply dts/bindings/rng/st,stm32-rng.yaml:20]: I saw that htcfg and htcfg1 are the register section names coming from the Ref Manual, 
I asked this change on purpose because: 
- When possible, using human readable names makes life easier to everyone. It could happen that sometime, register name is transparent and helps giving an idea of its content, but it's not the case here. You can see that in general, as a convention, device tree property names are trying to be self explanatory.
- "health test configuration" could also be found in the reference manual
- There is no guarantee that this name will persist across series and it might happen that for an upcoming series, a slightly different name is used. And then we would have a clear and unfortunate disconnection between device tree property and reference manual and all benefit lost.

**Review by mbolivar-nordic — APPROVED** [review]
Binding changes LGTM modulo any bikeshedding of the names, which is your call

**Comment by ceolin** [comment]
;;; Type C-c C-c to include your comment
Shouldn't this health test be done also when the SoC resumes from a sleep state ?

**Comment by ceolin** [comment]
;;; Type C-c C-c to include your comment
Shouldn't it be done when the SoC resumes from a sleep state ?

File: drivers/entropy/entropy_stm32.c
Comment (ceolin) [inline drivers/entropy/entropy_stm32.c:501]: Shouldn't it be done when the SoC resumes from a sleep state ?

File: dts/bindings/rng/st,stm32-rng.yaml
Comment (ABOSTM) [reply dts/bindings/rng/st,stm32-rng.yaml]: OK. Done.

File: dts/bindings/rng/st,stm32-rng.yaml
Comment (ABOSTM) [reply dts/bindings/rng/st,stm32-rng.yaml]: Done

File: drivers/entropy/entropy_stm32.c
Comment (ABOSTM) [reply drivers/entropy/entropy_stm32.c:501]: Nothing special to RNG. Like other peripherals, registers content is lost when wakeup from Standby/shutdown but in such case there is a full zephyr initialization is performed.

**Review by erwango — APPROVED** [review]

**Review by FRASTM — COMMENTED** [review]
some the STM32L4 devices also have a RNG Health Test Control
and a **LL_RNG_SetHealthconfiguration** function. 
Should this serie be included here too ?

File: dts/arm/st/h7/stm32h723.dtsi
Comment (FRASTM) [inline dts/arm/st/h7/stm32h723.dtsi:41]: or lower case ?

File: drivers/entropy/entropy_stm32.c
Comment (FRASTM) [inline drivers/entropy/entropy_stm32.c:491]: instead of relying on the DTS, checking the device and the LL_RNG_SetHealthConfig is defined
```
#ifdef (RNG_HTCR)
LL_RNG_SetHealthConfig
```

File: dts/arm/st/wl/stm32wl.dtsi
Comment (FRASTM) [inline dts/arm/st/wl/stm32wl.dtsi:410]: when device requires a  magic Nb its value is always 0x17590ABC, 
On the other hand it must not be omitted in the driver even if not present in the DTS.

Comment (ABOSTM) [reply dts/arm/st/wl/stm32wl.dtsi]: > some the STM32L4 devices also have a RNG Health Test Control and a **LL_RNG_SetHealthconfiguration** function. Should this serie be included here too ?

Yes, some L4 support this health test configuration, but only L4P and L4Q which are not yet supported in zephyr.

File: drivers/entropy/entropy_stm32.c
Comment (ABOSTM) [reply drivers/entropy/entropy_stm32.c:491]: This is not possible:
RNG_HTCR doesn't exist, but rather RNG->HTCR.
Some series define RNG_HTCR_HTCFG, but not all series, so it is not usable as generic mechanism.
And LL_RNG_SetHealthConfig() is a function not a macro, so it cannot be used for preprocessing.
So I rely on DTS. Up to user to set it correctly.

**Pull request merged**
